kornia/enhance/adjust.py

```diff
@@ -633,69 +633,116 @@ def _blend_one(input1: torch.Tensor, input2: torch.Tensor, factor: torch.Tensor)
     return torch.clamp(res, 0, 1)
 
 
-def _build_lut(histo, step):
-    # Compute the cumulative sum, shifting by step // 2
-    # and then normalization by step.
-    lut = (torch.cumsum(histo, 0) + (step // 2)) // step
-    # Shift lut, prepending with 0.
-    lut = torch.cat([torch.zeros(1, device=lut.device, dtype=lut.dtype), lut[:-1]])
-    # Clip the counts to be in range.  This is done
-    # in the C code for image.point.
-    return torch.clamp(lut, 0, 255)
-
-
-# Code taken from: https://github.com/pytorch/vision/pull/796
-def _scale_channel(im: torch.Tensor) -> torch.Tensor:
-    r"""Scale the data in the channel to implement equalize.
+def _equalize_channels(
+    im: torch.Tensor,
+    bins: Optional[int] = None,
+    minimum: Optional[Union[float, int]] = None,
+    maximum: Optional[Union[float, int]] = None,
+    dtype=None,
+) -> torch.Tensor:
+    if im.is_floating_point() and bins is None:
+        raise ValueError(f"`bins` are required for floating point inputs")
+
+    # infer the bins from the dynamic range of an integer tensor
+    im_min = im.min() if minimum is None else minimum
+    im_max = im.max() if maximum is None else maximum
+    im_bins = im_max - im_min + 1 if bins is None else bins
+
+    # convert to floating point and build a histogram
+    if dtype is None and not im.is_floating_point():
+        im = im.to(torch.float)
+    elif dtype is not None:
+        im = im.to(dtype)
+
+    hist = torch.histc(im, bins=im_bins, min=im_min, max=im_max)
+
+    # start from zero and drop the last element so we have a 0.5 mean
+    hist = torch.cat([torch.zeros(1, dtype=hist.dtype, device=hist.device), hist[:-1]])
+
+    # build a lookup table normalized to [0, 1]
+    lut = torch.cumsum(hist, 0) / hist.sum()
+
+    # expand to match the input tensor dims to avoid reshaping
+    lut = lut.expand(*im.shape[:-1] + lut.shape)
+
+    # gather along the channel dimension
+    return torch.gather(lut, im.ndim - 1, im - im_min)
+
+
+def _equalize_batch(
```

Comment:
All functional functions shall support only BCHW or BCDHW in current Kornia. Other formats are supported in nn.Module implementations.

Reply:
Ok, should I pass the channel dimension in instead and tranpose internally instead of doing it in equalize/equalize3d? I've been treating this as an implementation detail and not a public function

Reply:
I think it is better to treat it as an external function though you implemented as a private one~